Radio button errors

misaka opened the pull request.

aliuk2012 commented on Gemfile:
```diff
@@ -13,3 +13,5 @@ gemspec
 # gem 'byebug', group: [:development, :test]
 
+
+gem 'rspec-html-matchers', git: 'https://github.com/ministryofjustice/rspec-html-matchers.git', branch: 'fix-deep-nesting'
```
Once the rspec-html-matcher PR is merged we should remove the reference to the branch. Ideally, we should get push this up. We might want to mention in the forked repo that its just being used for our form builder so if someone looks at https://github.com/ministryofjustice/rspec-html-matchers.git and wonders what it's being used for and simply deletes the repo then it would break the tests.

aliuk2012 replied:
Should this maybe go in the gemspec as a dev dependency

misaka replied:
The gem itself is already in the `gemspec`, it was merged in with `cleanup-error-helpers-spec`. I don't believe we can specify a git repo or path with `add_development_dependency`, so this is in the Gemfile just for that purpose.

aliuk2012 commented:
👍

misaka commented:
@aliuk2012 I've cleaned up the commits here a bit ... and did some minor tweaks as I went. I removed the `ContactMethod` class since we don't use it in the tests anymore, and removed the use of the branch for `ministryofjustice/rspec-html-matchers`.

robmckinnon commented:
@misaka Thanks for contributing! The gem's used in multiple projects and departments. When html output is changed, I'm looking for:
* an explanation of what has changed;
* an explanation of why it has changed, see the [GDS git commit content guide](https://github.com/alphagov/styleguides/blob/master/git.md#content) for details; and
* an example added to the [examples website](https://govuk-elements-rails-guide.herokuapp.com), its repo is [ministryofjustice/govuk_elements_rails_guide](https://github.com/ministryofjustice/govuk_elements_rails_guide).

Are you attempting to replicate the [radio button error example html](http://govuk-elements.herokuapp.com/errors/#summarise-errors) from the GOV.UK elements guide? 

I'm finding it hard to review and agree what has changed as the [main commit](https://github.com/ministryofjustice/govuk_elements_form_builder/pull/59/commits/283582708f5704f516e65465cff954a126df6aca) both refactors specs and changes functionality. Are you able to split the code changes into:

1. a PR for radio button feature changes with existing specs updated but not refactored; and
1. a separate PR for the extensive refactoring of the specs?

We used a copy of HTML as spec expectations so far. We favoured having expectations that could be read and understood by those who know HTML and wanted to quickly compare to the GOV.UK elements guide markup examples. The tradeoff meant we have verbose expectations, with some repetition.

misaka commented:
@robmckinnon Thanks for chiming in. I've been working on this with @aliuk2012 and thought it was just us two paying attention.

You're completely right about the main commit needing some sorting out, I'll separate the two, possibly into separate PRs. Maybe @aliuk2012 can talk a bit better about why we're doing these changes now, but essentially yes, we needed to add the error reporting to the radio button, and noticed while we were doing it some discrepancies in the non-error HTML that we thought we'd address.

misaka commented:
Just to mention that I'm still reworking these commits a little bit to tidy them up. I'll be moving the refactoring commits to a separate PR.

misaka commented:
Ok, that happened a little faster than I thought it would. @robmckinnon I think this is ready for another review, and as I mentioned I'll put the test refactoring into another PR.

misaka commented:
I just realised I left the commit in that was pointing to our version of `rspec-html-matchers`. This isn't necessary with the spec refactoring removed so I've removed it from this PR.

robmckinnon commented:
The changes look good, thanks for reworking the commits!

Are you able to add a radio buttons error example added to the [examples website](https://govuk-elements-rails-guide.herokuapp.com)? Its repo is here: [ministryofjustice/govuk_elements_rails_guide](https://github.com/ministryofjustice/govuk_elements_rails_guide)

aliuk2012 commented:
@robmckinnon @misaka, I'm thinking that once this PR and #60 are merged we should make a release and put it up on rubygems so we can stats how many downloads etc this gem generates. I'm thinking release as version 1.0.0

robmckinnon commented:
A release to rubygems sounds fine. I suggest we update the [examples website](https://govuk-elements-rails-guide.herokuapp.com) and reference it from the gem docs.

A lower version number than 1.0.0 may allow people to be move forgiving of any issues that arise.

aliuk2012 commented:
@robmckinnon as the examples website depends on this gem, I'll merge it now and then work on the radio button error example. 

Good point about a lower version number for rubygems
